src/pool.ts

```diff
@@ -0,0 +1,677 @@
+/*!
+ * This source file is part of the EdgeDB open source project.
+ *
+ * Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import connect, {AwaitConnection, QueryArgs} from "./client";
+import {InvalidValueError} from "./errors";
+import {ConnectConfig} from "./con_utils";
+import {Set} from "./datatypes/set";
+import {LifoQueue} from "./queues";
+
+export class Deferred<T> {
+  private _promise: Promise<T>;
+  private resolve!: (value?: T | PromiseLike<T> | undefined) => void;
+  private reject!: (reason?: any) => void;
+  private _result: T | PromiseLike<T> | undefined;
+  private _done: boolean;
+
+  public get promise(): Promise<T> {
+    return this._promise;
+  }
+
+  public get done(): boolean {
+    return this._done;
+  }
+
+  public get result(): T | PromiseLike<T> | undefined {
+    if (!this._done) {
+      throw new InterfaceError("The deferred is not resolved");
+    }
+    return this._result;
+  }
+
+  setResult(value?: T | PromiseLike<T> | undefined): void {
+    this.resolve(value);
```

Comment:
`setTimeout` would be unnecessarily slow for this, use `process.nextTick`:

```
await new Promise((resolve) => process.nextTick(resolve));
```

> Normally, awaiting a timeout of 0 in a while loop would cause high
> CPU usage

Not really, it's not a busy loop, it's only executing once per event loop iteration.

I think the hack is OK for now. I plan to later add a proper implementation of `Deferred` that wouldn't rely on `Promises` at all. There's some code in `client.ts` that would be simplified if we had it.